src/offsets.rs

```diff
@@ -0,0 +1,237 @@
+/// The `txn` id for each partition is constructed as `<app_id>-<partition>` and used across
+/// kafka-delta-ingest to track messages offsets and protect from data duplication.
+///
+/// However, if table has already stored offset for given app_id/partition then this action could
+/// be ignored if stored offsets are equals or greater than provided seeking offsets.
+/// But, if stored offsets are lower then the `InconsistentStoredOffsets` is returned since it
+/// could break the data integrity.
+/// Hence, one is advised to supply the new `app_id` if skipping offsets is what required.
+pub(crate) async fn write_offsets_to_delta(
+    table: &mut DeltaTable,
+    app_id: &str,
+    offsets: &[(DataTypePartition, DataTypeOffset)],
+) -> Result<(), WriteOffsetsError> {
+    let offsets_as_str: String = offsets
+        .iter()
+        .map(|(p, o)| format!("{}:{}", p, o))
+        .collect::<Vec<String>>()
+        .join(",");
+
+    info!(
+        "Writing offsets [{}] to delta table {}",
+        offsets_as_str, table.table_uri
+    );
+
+    let mapped_offsets: Vec<(String, DataTypeOffset)> = offsets
+        .iter()
+        .map(|(p, o)| (txn_app_id_for_partition(app_id, *p), *o))
+        .collect();
+
+    if is_safe_to_commit_transactions(table, &mapped_offsets) {
+        // table has no stored offsets for given app_id/partitions so it is safe to write txn actions
+        commit_partition_offsets(table, mapped_offsets, &offsets_as_str).await?;
+        Ok(())
+    } else {
+        // there's at least one app_id/partition stored in delta,
+        // checking whether it's safe to proceed further
+        let mut conflict_offsets = Vec::new();
+
+        for (txn_app_id, offset) in mapped_offsets {
+            match table.get_app_transaction_version().get(&txn_app_id) {
+                Some(stored_offset) if *stored_offset < offset => {
+                    conflict_offsets.push((txn_app_id, *stored_offset, offset));
+                }
+                _ => (),
+            }
+        }
+
+        if conflict_offsets.is_empty() {
+            // there's no conflicted offsets in delta, e.g. it's either missing or is higher than seek offset
+            info!(
+                "The provided offsets are already applied for table {}",
+                table.table_uri
+            );
+            Ok(())
+        } else {
+            let partitions = conflict_offsets
+                .iter()
+                .map(|p| p.0.split('-').last().unwrap_or("N/A"))
+                .collect::<Vec<&str>>()
+                .join(",");
+
+            error!(
+                "Stored offsets for partitions [{}] are lower than provided offsets in table {}",
+                partitions, table.table_uri
+            );
+
+            let detailed_error_msg = conflict_offsets
+                .iter()
+                .map(|(partition, stored, provided)| {
+                    format!("{}:stored={}/provided={}", partition, stored, provided)
```

Comment:
And here as well

```suggestion
                    format!("{}:stored offset={}/seek offset={}", partition, stored, provided)
```

Reply:
do we want the offset ending each time? given there'll be a dozens of partition, I don't feel this is necessary

Reply:
btw, the reason why I didn't used seek here is that in this context, the seek doesn't make sense, as we just try to add a bunch of txn actions

Reply:
hence provided would make more sense, e.z. before vs after, stored vs provided

Reply:
anyway, left with _seek_

Reply:
this is a nitpick so i'm fine either way. i get what you're saying about the local function context not knowing that the "provided" offset happens to be a seek. the only place its used is in support of `seek-offsets` though so I think the "seek" terminology would make better sense to a user viewing the error message. i'm totes fine with either.